File: ArryLoad_VB/ArryLoad/NAudioRecorder.vb
Imports NAudio.Wave
Imports System
Imports System.Diagnostics
Imports System.Runtime.CompilerServices

Namespace ArryLoad
	Friend Class NAudioRecorder
		Implements ISpeechRecorder

		Private maxValue As Single

		Private minValue As Single

		Private count As Integer

		Private fftLength As Integer

		Private channels As Integer

		Public waveSource As WaveIn = Nothing

		Public waveFile As WaveFileWriter = Nothing

		Private fileName As String = String.Empty

		<CompilerGenerated()>
		<DebuggerBrowsable(DebuggerBrowsableState.Never), CompilerGenerated()>
		Public Event MaximumCalculated As EventHandler(Of MaxSampleEventArgs)

		<CompilerGenerated()>
		<DebuggerBrowsable(DebuggerBrowsableState.Never), CompilerGenerated()>
		Public Event DataAvailable As Action(Of Integer)

		Public Property NotificationCount() As Integer

		Public Property PerformFFT() As Boolean

		Public Sub New()
			Me.fftLength = 1024
			Me.channels = 2
			Dim flag As Boolean = Not NAudioRecorder.IsPowerOfTwo(Me.fftLength)
			If flag Then
				Throw New ArgumentException("FFT Length must be a power of two")
			End If
		End Sub

		Private Shared Function IsPowerOfTwo(x As Integer) As Boolean
			Return(x And x - 1) = 0
		End Function

		Public Sub StartRec()
			Me.waveSource = New WaveIn()
			Me.waveSource.WaveFormat = New WaveFormat(16000, 16, 2)
			AddHandler Me.waveSource.DataAvailable, AddressOf Me.waveSource_DataAvailable
			AddHandler Me.waveSource.RecordingStopped, AddressOf Me.waveSource_RecordingStopped
			Me.waveFile = New WaveFileWriter(Me.fileName, Me.waveSource.WaveFormat)
			Me.waveSource.StartRecording()
		End Sub

		Public Sub StopRec()
			Dim flag As Boolean = Me.waveSource IsNot Nothing
			If flag Then
				Me.waveSource.StopRecording()
			End If
		End Sub

		Public Sub SetFileName(fileName As String)
			Me.fileName = fileName
		End Sub

		Private Sub waveSource_DataAvailable(sender As Object, e As WaveInEventArgs)
			Dim flag As Boolean = Me.waveFile IsNot Nothing
			If flag Then
				Me.waveFile.Write(e.Buffer, 0, e.BytesRecorded)
				Me.waveFile.Flush()
				Dim obj As Integer = CInt((Me.waveFile.Length / CLng(Me.waveFile.WaveFormat.AverageBytesPerSecond)))
				Dim expr_5C As Action(Of Integer) = Me.DataAvailable
				If expr_5C IsNot Nothing Then
					expr_5C(obj)
				End If
				Dim array As Single() = New Single(e.Buffer.Length / Me.channels - 1) {}
				Dim num As Integer = 0
				For i As Integer = 0 To e.Buffer.Length - 1
					Dim arg_A0_0 As Single() = array
					Dim expr_88 As Integer = num
					num = expr_88 + 1
					arg_A0_0(expr_88) = CSng(BitConverter.ToInt16(e.Buffer, i)) / 32768F
				Next
				For j As Integer = 0 To array.Length - 1
					Me.Add(array(j))
				Next
			End If
		End Sub

		Private Sub Add(value As Single)
			Me.maxValue = Math.Max(Me.maxValue, value)
			Me.minValue = Math.Min(Me.minValue, value)
			Me.count += 1
			Dim flag As Boolean = Me.count >= Me.waveSource.WaveFormat.SampleRate / 100
			If flag Then
				Dim expr_5C As EventHandler(Of MaxSampleEventArgs) = Me.MaximumCalculated
				If expr_5C IsNot Nothing Then
					expr_5C(Me, New MaxSampleEventArgs(Me.minValue, Me.maxValue))
				End If
				Me.count = 0
				Me.maxValue = 0F
				Me.minValue = 0F
			End If
		End Sub

		Private Sub waveSource_RecordingStopped(sender As Object, e As StoppedEventArgs)
			Dim flag As Boolean = Me.waveSource IsNot Nothing
			If flag Then
				Me.waveSource.Dispose()
				Me.waveSource = Nothing
			End If
			Dim flag2 As Boolean = Me.waveFile IsNot Nothing
			If flag2 Then
				Me.waveFile.Dispose()
				Me.waveFile = Nothing
			End If
		End Sub

		Public Sub Reset()
			Me.count = 0
			Dim expr_0F As Single = 0F
			Dim num As Single = expr_0F
			Me.minValue = expr_0F
			Me.maxValue = num
		End Sub
	End Class
End Namespace

File: ArryLoad/NAudio/NAudioRecorder.cs
using NAudio.Dsp;
using NAudio.Wave;
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Timers;

namespace ArryLoad
{
    class NAudioRecorder : ISpeechRecorder
    {
        public event EventHandler<MaxSampleEventArgs> MaximumCalculated;
        private float maxValue;
        private float minValue;
        public int NotificationCount { get; set; }
        int count;

        public bool PerformFFT { get; set; }
        private readonly int fftLength;

        private readonly int channels;

        public WaveIn waveSource = null;
        public WaveFileWriter waveFile = null;
        private string fileName = string.Empty;
        public event Action<int> DataAvailable;

        public NAudioRecorder()
        {
            fftLength = 1024;
            channels = 2;
            if (!IsPowerOfTwo(fftLength))
            {
                throw new ArgumentException("FFT Length must be a power of two");
            }
        }

        static bool IsPowerOfTwo(int x)
        {
            return (x & (x - 1)) == 0;
        }

        /// <summary>
        /// 开始录音
        /// </summary>
        public void StartRec()
        {
            waveSource = new WaveIn();
            waveSource.WaveFormat = new WaveFormat(16000, 16, 2); // 16bit,16KHz,Mono的录音格式
            waveSource.DataAvailable += new EventHandler<WaveInEventArgs>(waveSource_DataAvailable);
            waveSource.RecordingStopped += new EventHandler<StoppedEventArgs>(waveSource_RecordingStopped);

            waveFile = new WaveFileWriter(fileName, waveSource.WaveFormat);
            waveSource.StartRecording();
        }

        /// <summary>
        /// 停止录音
        /// </summary>
        public void StopRec()
        {
            if (waveSource != null)
            {
                waveSource.StopRecording();
            }
        }

        /// <summary>
        /// 录音结束后保存的文件路径
        /// </summary>
        /// <param name="fileName">保存wav文件的路径名</param>
        public void SetFileName(string fileName)
        {
            this.fileName = fileName;
        }


        /// <summary>
        /// 开始录音回调函数
        /// </summary>
        /// <param name="sender"></param>
        /// <param name="e"></param>
        private void waveSource_DataAvailable(object sender, WaveInEventArgs e)
        {
            if (waveFile != null)
            {
                waveFile.Write(e.Buffer, 0, e.BytesRecorded);
                waveFile.Flush();

                int secondsRecorded = (int)(waveFile.Length / waveFile.WaveFormat.AverageBytesPerSecond);
                DataAvailable?.Invoke(secondsRecorded);

                float[] sts = new float[e.Buffer.Length / channels];
                int outIndex = 0;
                for (int n = 0; n < e.Buffer.Length; n += channels)
                {
                    sts[outIndex++] = BitConverter.ToInt16(e.Buffer, n) / 32768f;
                }

                for (int n = 0; n < sts.Length; n += channels)
                {
                    Add(sts[n]);
                }
            }
        }

        private void Add(float value)
        {
            maxValue = Math.Max(maxValue, value);
            minValue = Math.Min(minValue, value);
            count++;
            if (count >= waveSource.WaveFormat.SampleRate / 100)
            {
                MaximumCalculated?.Invoke(this, new MaxSampleEventArgs(minValue, maxValue));
                count = 0;
                maxValue = 0;
                minValue = 0;
            }
        }



        /// <summary>
        /// 录音结束回调函数
        /// </summary>
        /// <param name="sender"></param>
        /// <param name="e"></param>
        private void waveSource_RecordingStopped(object sender, StoppedEventArgs e)
        {
            if (waveSource != null)
            {
                waveSource.Dispose();
                waveSource = null;
            }

            if (waveFile != null)
            {
                waveFile.Dispose();
                waveFile = null;
            }
        }

        public void Reset()
        {
            count = 0;
            maxValue = minValue = 0;
        }
    }

    public class MaxSampleEventArgs : EventArgs
    {
        [DebuggerStepThrough]
        public MaxSampleEventArgs(float minValue, float maxValue)
        {
            MaxSample = maxValue;
            MinSample = minValue;
        }
        public float MaxSample { get; private set; }
        public float MinSample { get; private set; }
    }

    public class FftEventArgs : EventArgs
    {
        [DebuggerStepThrough]
        public FftEventArgs(Complex[] result)
        {
            Result = result;
        }
        public Complex[] Result { get; private set; }
    }
}
